src/qutip_qip/decompositions/general_decompositions.py

```diff
@@ -0,0 +1,113 @@
+    if qobj_check == False:
+        raise TypeError("The input matrix is not a Qobj.")
+    else:
+        # check if input is square and a unitary
+        input_shape = input_gate.shape
+        # check if input is 1 row and 1 column matrix
+        if input_shape[0]==1:
+            raise ValueError("A 1-D Qobj is not a valid quantum gate.")
+        # check if the input is a rectangular matrix
+        if input_shape[0] != input_shape[1]:
+            raise ValueError("Input is not a square matrix.")
+        else:
+            unitary_check = Qobj.check_isunitary(input_gate)
+            return unitary_check
+
+
+def check_input_shape(input_gate, num_of_qubits):
+    """Check if the shape of input gate is valid to act on input number of qubits
+
+    If :math:`n` is the number of qubits in the circuit then a valid quantum gate
+    acting on these qubits must be of dimension :math:`2^n \\times 2^n`.
+
+    Parameters
+    ----------
+    input_gate : :class:`qutip.Qobj`
+        The matrix that's supposed to be decomposed should be a Qobj.
+
+    num_of_qubits : int
+        Number of qubits in the circuit.
+
+    Returns
+    -------
+    bool
+        Returns "True" if the shape of input gate is valid to act on the number
+        of qubits in the circuit.
+
+    """
+    input_check_bool = check_input(input_gate)
+    if input_check_bool == True:
+        input_shape = input_gate.shape
+        try:
+            assert input_shape[0] == 2**num_of_qubits
+        except AssertionError:
+            return(False)
+        return(True)
+    else:
+        raise ValueError("Input is not unitary.")
+
+
+def convert_qobj_gate_to_array(input_gate):
+    """Converts a valid unitary quantum gate to a numpy array.
+
+    Parameters
+    ----------
+    input_gate : :class:`qutip.Qobj`
+        The matrix that's supposed to be decomposed should be a Qobj.
+
+    Returns
+    -------
+    input_gate : `np.array`
+        The input is returned as a converted numpy array.
+    """
+    input_check_bool = check_input(input_gate)
+    if input_check_bool == True:
+        input_to_array = Qobj.full(input_gate)
+        try:
+            isinstance(input_to_array, np.ndarray)
+        except:
+            raise ConversionError("Input Qobj could not be converted to a numpy array.")
+        return(input_to_array)
+    else:
+        raise ValueError("Input is not unitary.")
+
+
+def extract_global_phase(input_gate, num_of_qubits):
+    """ Extracts some common constant from all the elements of the matrix. The output
+    is retuned in the complex exponential form.
+
+    Parameters
+    ----------
+    input_gate : :class:`qutip.Qobj`
+        The matrix that's supposed to be decomposed should be a Qobj.
+
+    """
+    input_array = convert_qobj_gate_to_array(input_gate)
+    determinant_of_input = np.linalg.det(input_array)
+    y = np.imag(determinant_of_input)
+    x = np.real(determinant_of_input)
+    global_phase_angle = np.arctan2(y,x)
```

Comment:
Is it possible that the problem is here? Since I remember you mentioned a few times about using `cmath.phase` instead of `arctan`.

Reply:
Not for `arctan2`. `arctan` and `arctan2` are two different functions in numpy. 

I had made a mistake by trying to use `arctan`. 

I think the issue with some of the outputs being off by some global phase could be the lack of a normalization constant (or division by absolute value). I did think of this initially but since I check if the matrix is unitary before extracting the phase, I didn't think I had to have it.